concordium-consensus/src/Concordium/ImportExport.hs

```diff
@@ -127,76 +226,92 @@ writeSection ::
+    -- |Export directory
+    FilePath ->
+    -- |Index file handle
     Handle ->
-    -- |Action to write out the blocks and return the number of them
-    m Word64 ->
-    -- |Action to write out the finalization records and return the number of them
-    m Word64 ->
+    -- |Chunk size in blocks
+    Word64 ->
     m ()
-writeSection
+writeChunks
     sectionGenesisIndex
     sectionProtocolVersion
     sectionGenesisHash
     sectionFirstBlockHeight
-    hdl
-    writeBlocks
-    writeFinRecs = do
+    sectionLastBlockHeight
+    outDir indexHdl chunkSize = do
+        let chunkName = outDir </> "blocks-"
+                        ++ show sectionGenesisIndex ++ "-"
+                        ++ show sectionFirstBlockHeight
+                        ++ ".dat"
+        chunkHdl <- liftIO $ initialHandle chunkName
+
         (sectionStart, blocksStart) <- liftIO $ do
-            sectionStart <- hTell hdl
+            BS.hPut chunkHdl $ encode (3 :: Version)
+            sectionStart <- hTell chunkHdl
             -- Write a dummy section header that we will later overwrite
-            runPutH (liftPut $ putWord64be sectionHeaderLength >> put placeholderSectionHeader) hdl
-            blocksStart <- hTell hdl
+            runPutH (liftPut $ putWord64be sectionHeaderLength >> put placeholderSectionHeader) chunkHdl
+            blocksStart <- hTell chunkHdl
             return (sectionStart, blocksStart)
-        sectionBlockCount <- writeBlocks
-        blocksEnd <- liftIO $ hTell hdl
-        sectionFinalizationCount <- writeFinRecs
+        sectionBlockCount <- exportBlocksToChunk chunkHdl sectionFirstBlockHeight chunkSize
+        blocksEnd <- liftIO $ hTell chunkHdl
+        -- Only write finalization records to a chunk if it's the last one for the section
+        let lastExportedBlockHeight = sectionFirstBlockHeight + BlockHeight sectionBlockCount - 1
+        sectionFinalizationCount <- if lastExportedBlockHeight < sectionLastBlockHeight
+                                    then return 0
+                                    else exportFinRecsForChunk chunkHdl
         liftIO $ do
-            sectionEnd <- hTell hdl
+            sectionEnd <- hTell chunkHdl
             -- Go back to the start and rewrite the section header with the correct data
-            hSeek hdl AbsoluteSeek sectionStart
+            hSeek chunkHdl AbsoluteSeek sectionStart
             let sectionHeader =
                     SectionHeader
                         { sectionLength = fromInteger (sectionEnd - sectionStart),
                           sectionBlocksLength = fromInteger (blocksEnd - blocksStart),
                           ..
                         }
-            runPutH (liftPut $ putWord64be sectionHeaderLength >> put sectionHeader) hdl
-            hSeek hdl AbsoluteSeek sectionEnd
+            runPutH (liftPut $ putWord64be sectionHeaderLength >> put sectionHeader) chunkHdl
+            hClose chunkHdl
+            TIO.hPutStrLn indexHdl . T.pack $
```

Comment:
What is the benefit of using `TIO.hPutStrLn` here, rather than `System.IO.hPutStrLn`? I think since we're using `String`s already there's nothing gained.